File: backend/app/api/influencers.py
from fastapi import APIRouter, UploadFile, File, HTTPException, status
import pandas as pd
import io
import uuid
from datetime import datetime
from app.core.config import db
from google.cloud import firestore
import math

router = APIRouter()

REQUIRED_COLUMNS = [
    "PROFILE LINK", "NAME", "LOCATION", "NICHE"
]

def clean_float(value):
    """Safely convert value to float, handling diverse formats."""
    if pd.isna(value) or value == "":
        return 0.0
    if isinstance(value, (int, float)):
        return float(value)
    if isinstance(value, str):
        # Remove commas, percentage signs, etc.
        cleaned = value.replace(",", "").replace("%", "").strip()
        try:
            return float(cleaned)
        except ValueError:
            return 0.0
    return 0.0

def clean_str(value):
    """Safely clean string values."""
    if pd.isna(value) or value is None:
        return ""
    return str(value).strip()

@router.post("/upload-csv", status_code=status.HTTP_201_CREATED)
async def upload_csv(file: UploadFile = File(...)):
    """
    Uploads a CSV file containing influencer data and stores it in Firestore.
    """
    if not file.filename.endswith('.csv'):
        raise HTTPException(status_code=400, detail="Invalid file type. Please upload a CSV file.")

    try:
        contents = await file.read()
        # Decode contents to handle various encodings
        content_str = contents.decode('utf-8-sig') # handle BOM if present
        
        # Try different delimiters
        def try_parse(sep):
            return pd.read_csv(io.StringIO(content_str), sep=sep)

        try:
            # First try default comma
            df = try_parse(',')
            # If only one column, try tab (common when copy-pasting from Sheets/Excel)
            if len(df.columns) <= 1:
                df = try_parse('\t')
        except:
             # Fallback
             df = pd.read_csv(io.StringIO(content_str))

        # Normalize column names: strip whitespace and convert to upper case
        # This fixes issues where " NAME" or "name" wouldn't match "NAME"
        df.columns = [col.strip().upper() for col in df.columns]
        print(f"Detected columns: {df.columns.tolist()}")

    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Error parsing CSV: {str(e)}")

    # Validate Columns (using normalized list)
    # Check if we have at least the critical ones
    available_cols = df.columns.tolist()
    missing_cols = [col for col in REQUIRED_COLUMNS if col not in available_cols]
    if missing_cols:
        # If headers are totally weird, let's provide more info
        raise HTTPException(status_code=400, detail=f"Missing required columns: {', '.join(missing_cols)}. Available: {', '.join(available_cols[:5])}...")

    created_at = datetime.utcnow()
    campaign_id = "test_campaign_001"
    collection_ref = db.collection("campaigns").document(campaign_id).collection("influencers")
    
    count = 0
    batch = db.batch()
    
    try:
        # Iterate through rows
        for _, row in df.iterrows():
            influencer_id = str(uuid.uuid4())
            
            # safely access columns (keys are now normalized upper case)
            def get_val(col):
                # Ensure we also search with upper/stripped
                return row.get(col, None)

            record = {
                "profile": {
                    "link": clean_str(get_val("PROFILE LINK")),
                    "name": clean_str(get_val("NAME")),
                    "gender": clean_str(get_val("GENDER")),
                    "location": clean_str(get_val("LOCATION")),
                    "type": clean_str(get_val("TYPE")),
                },
                "metrics": {
                    "followers": clean_float(get_val("FOLLOWERS")),
                    "avg_views": clean_float(get_val("AVERAGE VIEWS")),
                    "engagement_rate": clean_float(get_val("ENGAGEMENT RATE"))
                },
                "audience": {
                    "mf_split": clean_str(get_val("M/F SPLIT")),
                    "india_split": clean_str(get_val("INDIA 1/2 SPLIT")),
                    "age_concentration": clean_str(get_val("AGE CONCENTRATION"))
                },
                "brand": {
                    "niche": clean_str(get_val("NICHE")),
                    "brand_fit": clean_str(get_val("BRAND FIT")),
                    "vibe": clean_str(get_val("VIBE"))
                },
                "commercials": clean_str(get_val("COMMERCIALS")),
                "contact": {
                    "contact_no": clean_str(get_val("CONTACT NO.")),
                    "email": clean_str(get_val("EMAIL"))
                },
                "created_at": created_at
            }

            doc_ref = collection_ref.document(influencer_id)
            batch.set(doc_ref, record)
            count += 1
            
            # Firestore batch limit is 500
            if count % 400 == 0:
                batch.commit()
                batch = db.batch()

        # Commit remaining
        if count > 0:
            batch.commit()

    except Exception as e:
        print(f"Firestore Error: {str(e)}")
        if "firestore.googleapis.com" in str(e).lower():
            raise HTTPException(
                status_code=500, 
                detail="Firestore API is disabled. Please enable it in Google Cloud Console."
            )
        raise HTTPException(status_code=500, detail=f"Database error: {str(e)}")

    return {
        "status": "success",
        "uploaded_records": count,
        "campaign_id": campaign_id
    }

@router.get("/influencers")
async def get_influencers(page: int = 1, page_size: int = 10, campaign_id: str = "test_campaign_001"):
    """
    Fetches influencers from Firestore with pagination.
    """
    try:
        if page < 1:
            page = 1
        
        collection_ref = db.collection("campaigns").document(campaign_id).collection("influencers")
        
        # Get total count (using aggregation for better performance if possible, but standard count for now)
        # Note: In production with large datasets, consider maintaining a counter or using Firestore count queries
        total_docs = collection_ref.count().get()[0][0].value
        
        # Query for paginated results
        # Firestore pagination works with cursor-based pagination (startAfter) or offset (limit/offset)
        # Offset is easier but less efficient for very large datasets. 
        # For simplicity and given the usage, we'll use limit/offset or just limit with ordering.
        
        query = collection_ref.order_by("created_at", direction=firestore.Query.DESCENDING).limit(page_size).offset((page - 1) * page_size)
        docs = query.stream()
        
        influencers = []
        for doc in docs:
            influencer_data = doc.to_dict()
            influencer_data["id"] = doc.id
            # Convert datetime to string for JSON serialization
            if "created_at" in influencer_data and isinstance(influencer_data["created_at"], datetime):
                influencer_data["created_at"] = influencer_data["created_at"].isoformat()
            influencers.append(influencer_data)
            
        return {
            "influencers": influencers,
            "total": total_docs,
            "page": page,
            "page_size": page_size,
            "total_pages": math.ceil(total_docs / page_size) if total_docs > 0 else 0
        }
        
    except Exception as e:
        print(f"Error fetching influencers: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Database error: {str(e)}")

File: server/src/api/influencers.js
import { Router } from "express";
import multer from "multer";
import { parse } from "csv-parse/sync";
import { v4 as uuidv4 } from "uuid";
import { db } from "../core/config.js";
import { firebaseAdmin } from "../core/config.js";

const router = Router();
const FieldValue = firebaseAdmin.firestore.FieldValue;

// Multer – accept CSV up to 10 MB in memory
const upload = multer({
    storage: multer.memoryStorage(),
    limits: { fileSize: 10 * 1024 * 1024 },
    fileFilter: (_req, file, cb) => {
        if (
            file.mimetype === "text/csv" ||
            file.originalname.toLowerCase().endsWith(".csv")
        ) {
            cb(null, true);
        } else {
            cb(new Error("Invalid file type. Please upload a CSV file."));
        }
    },
});

const REQUIRED_COLUMNS = ["PROFILE LINK", "NAME", "LOCATION", "NICHE"];

// ─── helpers ─────────────────────────────────────────────────────────
function cleanFloat(value) {
    if (value == null || value === "") return 0.0;
    if (typeof value === "number") return value;
    const cleaned = String(value).replace(/,/g, "").replace(/%/g, "").trim();
    const n = parseFloat(cleaned);
    return Number.isNaN(n) ? 0.0 : n;
}

function cleanStr(value) {
    if (value == null) return "";
    return String(value).trim();
}

// ─── POST /upload-csv ───────────────────────────────────────────────
router.post("/upload-csv", upload.single("file"), async (req, res) => {
    try {
        if (!req.file) {
            return res.status(400).json({ detail: "No file uploaded." });
        }

        const content = req.file.buffer.toString("utf-8").replace(/^\uFEFF/, ""); // strip BOM

        const records = parse(content, {
            columns: true,
            skip_empty_lines: true,
            trim: true,
            relax_column_count: true,
        });

        if (records.length === 0) {
            return res.status(400).json({ detail: "CSV file appears to be empty or missing headers." });
        }

        // Normalize column names to UPPER CASE
        const normalizedRecords = records.map((row) => {
            const out = {};
            for (const [key, val] of Object.entries(row)) {
                out[key.trim().toUpperCase()] = val;
            }
            return out;
        });

        const availableCols = Object.keys(normalizedRecords[0]);
        const missingCols = REQUIRED_COLUMNS.filter((c) => !availableCols.includes(c));
        if (missingCols.length) {
            return res.status(400).json({
                detail: `Missing required columns: ${missingCols.join(", ")}. Available: ${availableCols.slice(0, 5).join(", ")}...`,
            });
        }

        const createdAt = new Date();
        const campaignId = "test_campaign_001";
        const collectionRef = db.collection("campaigns").document
            ? db.collection("campaigns").doc(campaignId).collection("influencers")
            : db.collection("campaigns").doc(campaignId).collection("influencers");

        let batch = db.batch();
        let count = 0;

        for (const row of normalizedRecords) {
            const g = (col) => row[col] ?? null;
            const influencerId = uuidv4();

            const record = {
                profile: {
                    link: cleanStr(g("PROFILE LINK")),
                    name: cleanStr(g("NAME")),
                    gender: cleanStr(g("GENDER")),
                    location: cleanStr(g("LOCATION")),
                    type: cleanStr(g("TYPE")),
                },
                metrics: {
                    followers: cleanFloat(g("FOLLOWERS")),
                    avg_views: cleanFloat(g("AVERAGE VIEWS")),
                    engagement_rate: cleanFloat(g("ENGAGEMENT RATE")),
                },
                audience: {
                    mf_split: cleanStr(g("M/F SPLIT")),
                    india_split: cleanStr(g("INDIA 1/2 SPLIT")),
                    age_concentration: cleanStr(g("AGE CONCENTRATION")),
                },
                brand: {
                    niche: cleanStr(g("NICHE")),
                    brand_fit: cleanStr(g("BRAND FIT")),
                    vibe: cleanStr(g("VIBE")),
                },
                commercials: cleanStr(g("COMMERCIALS")),
                contact: {
                    contact_no: cleanStr(g("CONTACT NO.")),
                    email: cleanStr(g("EMAIL")),
                },
                created_at: createdAt,
            };

            const docRef = db
                .collection("campaigns")
                .doc(campaignId)
                .collection("influencers")
                .doc(influencerId);

            batch.set(docRef, record);
            count++;

            // Firestore batch limit is 500
            if (count % 400 === 0) {
                await batch.commit();
                batch = db.batch();
            }
        }

        if (count % 400 !== 0) {
            await batch.commit();
        }

        return res.status(201).json({
            status: "success",
            uploaded_records: count,
            campaign_id: campaignId,
        });
    } catch (err) {
        console.error("CSV upload error:", err);
        if (err.message?.includes("firestore.googleapis.com")) {
            return res.status(500).json({
                detail: "Firestore API is disabled. Please enable it in Google Cloud Console.",
            });
        }
        return res.status(500).json({ detail: `Database error: ${err.message}` });
    }
});

// ─── GET /influencers ───────────────────────────────────────────────
router.get("/influencers", async (req, res) => {
    try {
        let { page = 1, page_size = 10, search, campaign_id = "test_campaign_001" } = req.query;
        page = Math.max(1, parseInt(page, 10) || 1);
        page_size = Math.min(100, Math.max(1, parseInt(page_size, 10) || 10));

        const collectionRef = db
            .collection("campaigns")
            .doc(campaign_id)
            .collection("influencers");

        if (search) {
            const lowerSearch = search.toLowerCase();

            const allSnap = await collectionRef.orderBy("created_at", "desc").get();

            const filtered = [];
            allSnap.forEach((doc) => {
                const data = doc.data();
                const profile = data.profile || {};
                const contact = data.contact || {};

                const matches =
                    (profile.link || "").toLowerCase().includes(lowerSearch) ||
                    (profile.name || "").toLowerCase().includes(lowerSearch) ||
                    (profile.gender || "").toLowerCase().includes(lowerSearch) ||
                    (profile.location || "").toLowerCase().includes(lowerSearch) ||
                    (contact.contact_no || "").toLowerCase().includes(lowerSearch) ||
                    (contact.email || "").toLowerCase().includes(lowerSearch);

                if (matches) {
                    const item = { ...data, id: doc.id };
                    if (item.created_at && item.created_at.toDate) {
                        item.created_at = item.created_at.toDate().toISOString();
                    }
                    filtered.push(item);
                }
            });

            const total = filtered.length;
            const start = (page - 1) * page_size;
            const paginated = filtered.slice(start, start + page_size);

            return res.json({
                influencers: paginated,
                total,
                page,
                page_size,
                total_pages: total > 0 ? Math.ceil(total / page_size) : 0,
            });
        }

        // Non-search: paginated query
        const countSnap = await collectionRef.count().get();
        const totalDocs = countSnap.data().count;

        const snap = await collectionRef
            .orderBy("created_at", "desc")
            .limit(page_size)
            .offset((page - 1) * page_size)
            .get();

        const influencers = [];
        snap.forEach((doc) => {
            const data = doc.data();
            data.id = doc.id;
            if (data.created_at && data.created_at.toDate) {
                data.created_at = data.created_at.toDate().toISOString();
            }
            influencers.push(data);
        });

        return res.json({
            influencers,
            total: totalDocs,
            page,
            page_size,
            total_pages: totalDocs > 0 ? Math.ceil(totalDocs / page_size) : 0,
        });
    } catch (err) {
        console.error("Error fetching influencers:", err);
        return res.status(500).json({ detail: `Database error: ${err.message}` });
    }
});

export default router;
